internal/controller/stages/regular_stages.go

```diff
@@ -0,0 +1,1981 @@
+					break
+				}
+			}
+
+			if promo.Status.Phase.IsTerminal() {
+				info := kargoapi.PromotionReference{
+					Name:       promo.Name,
+					Status:     promo.Status.DeepCopy(),
+					FinishedAt: promo.Status.FinishedAt,
+				}
+				if promo.Status.Freight != nil {
+					info.Freight = promo.Status.Freight.DeepCopy()
+				}
+				newPromotions = append(newPromotions, info)
+			}
+		}
+
+		// As we will be appending to the Freight history, we need to ensure that
+		// we order the Promotions from oldest to newest. This is because the
+		// Freight history is garbage collected based on the number of entries,
+		// and we want to ensure that the oldest entries are removed first.
+		slices.SortFunc(newPromotions, func(a, b kargoapi.PromotionReference) int {
+			return strings.Compare(a.Name, b.Name)
+		})
+
+		// Update the Stage status with the information about the newly terminated
+		// Promotions, and any new Freight that was successfully promoted.
+		for _, p := range newPromotions {
+			promo := p
+			newStatus.LastPromotion = &promo
+			if p.Status.Phase == kargoapi.PromotionPhaseSucceeded {
+				// If the Promotion was successful, then we should add the Freight
+				// to the history of successfully promoted Freight.
+				newStatus.FreightHistory.Record(promo.Status.FreightCollection)
+
+				// Erase any health checks that were performed for the previous
+				// Freight, as they are no longer relevant.
+				newStatus.Health = nil
+				conditions.Set(&newStatus, &metav1.Condition{
+					Type:               kargoapi.ConditionTypeHealthy,
+					Status:             metav1.ConditionUnknown,
+					Reason:             "WaitingForHealthCheck",
+					Message:            "Waiting for health check to be performed after successful promotion",
+					ObservedGeneration: stage.Generation,
+				})
+
+				// Set verified condition to unknown to indicate that the
+				// new Freight needs to be verified.
+				conditions.Set(&newStatus, &metav1.Condition{
+					Type:               kargoapi.ConditionTypeVerified,
+					Status:             metav1.ConditionUnknown,
+					Reason:             "WaitingForVerification",
+					Message:            "Waiting for verification to be performed after successful promotion",
+					ObservedGeneration: stage.Generation,
+				})
+			}
+		}
+
+		// Return at this point to allow the new Freight to be verified.
+		return newStatus, hasNonTerminalPromotions, nil
+	}
+
+	// If the current Freight exists and has a non-terminal verification, wait
+	// for it to complete regardless of health state to ensure we capture the
+	// results.
+	if curFreight := stage.Status.FreightHistory.Current(); curFreight != nil {
+		if curFreight.HasNonTerminalVerification() {
+			logger.Debug(
+				"current Freight has a non-terminal verification: " +
+					"wait for it to complete before allowing new promotions to start",
+			)
+			conditions.Delete(&newStatus, kargoapi.ConditionTypePromoting)
+			return newStatus, hasNonTerminalPromotions, nil
```

Comment:
Ref: [this comment](https://github.com/akuity/kargo/pull/2905/files#r1842750619)